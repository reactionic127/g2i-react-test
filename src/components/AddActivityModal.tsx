import React, { ReactElement, useState } from "react";
import { useDispatch } from "react-redux";
import { Input, Modal, Select, Space } from "antd";
import { ModalType, SelectType } from "../types";
import { ADD_ACTIVITY_REQUEST } from "../redux/constants";

const { Option } = Select;
const { TextArea } = Input;

function AddActivityModal({ visible, setVisible }: ModalType): ReactElement {
  const dispatch = useDispatch();

  const [type, setType] = useState<string>("note");
  const [url, setUrl] = useState<string>("");
  const [note, setNote] = useState<string>("");

  const handleChange = (value: SelectType) => {
    setType(value.value);
  };

  const submitValues = () => {
    if (note) {
      dispatch({ type: ADD_ACTIVITY_REQUEST, payload: { type, url, note } });
      setVisible(false);
      setNote("");
      setUrl("");
    }
  };

  return (
    <Modal
      data-id="modal-container"
      title="Add Activity"
      centered
      visible={visible}
      okText="Submit"
      onOk={submitValues}
      cancelButtonProps={{ style: { display: "none" } }}
      onCancel={() => setVisible(false)}
    >
      <Space direction="vertical" size="large" style={{ width: "100%" }}>
        <Select
          data-id="select-type"
          labelInValue
          defaultValue={{ value: "note", label: "Add Note", key: "note" }}
          style={{ width: 150 }}
          onChange={handleChange}
        >
          <Option value="note">Add Note</Option>
          <Option value="enquiry">Add Enquiry</Option>
          <Option value="task">Add Task</Option>
        </Select>
        <Input
          data-id="property-input"
          placeholder="Add Property URL (optional)"
          style={{ width: 300 }}
          value={url}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setUrl(e.target.value)
          }
        />
        <TextArea
          data-id="note-input"
          rows={4}
          placeholder="Add notes..."
          style={{ width: "100%", resize: "none" }}
          value={note}
          onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) =>
            setNote(e.target.value)
          }
        />
      </Space>
    </Modal>
  );
}

export default AddActivityModal;
