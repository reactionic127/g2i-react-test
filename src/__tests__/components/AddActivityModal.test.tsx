import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import AddActivityModal from "../../components/AddActivityModal";

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: () => ({ activities: [] }),
  useDispatch: () => mockDispatch,
}));

describe("AddActivityModal", () => {
  describe("rendering", () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
      const visible = true;
      const setVisible = jest.fn;
      wrapper = shallow(
        <AddActivityModal visible={visible} setVisible={setVisible} />
      );
    });

    it("renders the component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should have Modal container", () => {
      const modal = wrapper.find('[data-id="modal-container"]');
      expect(modal.length).toEqual(1);
    });

    it("should have Select Element", () => {
      expect(wrapper.find('[data-id="select-type"]').length).toEqual(1);
    });

    it("should have Property URL Input", () => {
      expect(wrapper.find('[data-id="property-input"]').length).toEqual(1);
    });

    it("should have Notes Input", () => {
      expect(wrapper.find('[data-id="note-input"]').length).toEqual(1);
    });

    it("simulate onOk click event on Modal", () => {
      const modal = wrapper.find('[data-id="modal-container"]');
      expect(modal.simulate("ok")).toHaveLength(1);
    });

    it("simulate onCancel click event on Modal", () => {
      const modal = wrapper.find('[data-id="modal-container"]');
      expect(modal.simulate("cancel")).toHaveLength(1);
    });

    it("simulate Select onChange event", () => {
      const select = wrapper.find('[data-id="select-type"]');
      expect(
        select.simulate("change", {
          value: { name: "type", value: "type value" },
        })
      ).toHaveLength(1);
    });

    it("simulate Propery URL Input onChange event", () => {
      const propertyInput = wrapper.find('[data-id="property-input"]');
      expect(
        propertyInput.simulate("change", {
          target: { name: "username", value: "correctUsername" },
        })
      ).toHaveLength(1);
    });

    it("simulate Note Input onChange event", () => {
      const noteInput = wrapper.find('[data-id="note-input"]');
      expect(
        noteInput.simulate("change", {
          target: { name: "username", value: "correctUsername" },
        })
      ).toHaveLength(1);
    });
  });
});
