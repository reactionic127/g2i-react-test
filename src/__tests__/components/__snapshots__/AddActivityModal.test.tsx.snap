// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddActivityModal rendering renders the component 1`] = `
<Modal
  cancelButtonProps={
    Object {
      "style": Object {
        "display": "none",
      },
    }
  }
  centered={true}
  confirmLoading={false}
  data-id="modal-container"
  maskTransitionName="fade"
  okText="Submit"
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Add Activity"
  transitionName="zoom"
  visible={true}
  width={520}
>
  <Space
    direction="vertical"
    size="large"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <ForwardRef(InternalSelect)
      data-id="select-type"
      defaultValue={
        Object {
          "key": "note",
          "label": "Add Note",
          "value": "note",
        }
      }
      labelInValue={true}
      onChange={[Function]}
      style={
        Object {
          "width": 150,
        }
      }
    >
      <Option
        value="note"
      >
        Add Note
      </Option>
      <Option
        value="enquiry"
      >
        Add Enquiry
      </Option>
      <Option
        value="task"
      >
        Add Task
      </Option>
    </ForwardRef(InternalSelect)>
    <Input
      data-id="property-input"
      onChange={[Function]}
      placeholder="Add Property URL (optional)"
      style={
        Object {
          "width": 300,
        }
      }
      type="text"
      value=""
    />
    <ForwardRef
      data-id="note-input"
      onChange={[Function]}
      placeholder="Add notes..."
      rows={4}
      style={
        Object {
          "resize": "none",
          "width": "100%",
        }
      }
      value=""
    />
  </Space>
</Modal>
`;
